Allow a driver to customize the query remarks

robdaemon opened the pull request:
This allows driver-specific metadata to be added to each query. If the
driver can apply the default metadata, it should call
default-query->remark

Commit:
Allow a driver to customize the query remarks

This allows driver-specific metadata to be added to each query. If the
driver can apply the default metadata, it should call
default-query->remark

Commit:
Fixing docstring lint failure

Commit:
Fixing broken drivers due to signature change

Commit:
Adding optional name field to value literals

This will allow access to the field's name in the database down in the
drivers, without the drivers having to call back up into the schema
store.

Commit:
Test to assert that query remark is proper

Adds a test that the query remark is properly modified. Will be changing
as more fields are wired into that remark.

Commit:
Adding filter_values to the resulting remark

Traverses the :filter clause of a MBQL query to output the filter values
to the query remark

Commit:
Handle multiple values for the same field

When the same field is used in an AND/OR setting, show all values in the
field_values metadata

camsaul commented on modules/drivers/bigquery/src/metabase/driver/bigquery.clj:
```diff
@@ -228,7 +228,7 @@
     (binding [bigquery.common/*bigquery-timezone-id* (effective-query-timezone-id database)]
       (log/tracef "Running BigQuery query in %s timezone" bigquery.common/*bigquery-timezone-id*)
-      (let [sql (str "-- " (qputil/query->remark outer-query) "\n" (if (seq params)
+      (let [sql (str "-- " (qputil/query->remark :bigquery outer-query) "\n" (if (seq params)
```
might want to reindent the lines below

camsaul commented on src/metabase/query_processor/util.clj:
```diff
@@ -37,6 +37,18 @@
                             (codecs/bytes->hex query-hash)))))
 
+(defmulti query->remark
+  "Generate an appropriate REMARK to be prepended to a query to give DBAs additional information about the query being
```
Some of our really old code uses all uppercase for parameter names, but for new code you should follow the [Clojure Style Guide](https://guide.clojure.style/) and [use backticks instead](https://guide.clojure.style/#document-pos-arguments).

When we first started writing the Clojure version of Metabase there wasn't really any consensus on highlighting arg names in docstrings yet so I just followed the convention from Emacs Lisp and wrote a bit of Emacs Lisp to highlight them for me. These days pretty much everyone does backticks and my custom Emacs Lisp is long-gone (`clojure-mode`/CIDER highlights backticks automatically now).

My general strategy has been to update the all-uppercase names when I come across them while working on other stuff. Doing it all at once wasn't really appealing. It's been an ongoing process for a few years but I'd say 95% of our code uses backticks now.

So feel free to fix those when you see them if you want. If you don't, it doesn't really make a huge difference to me either. Just make sure not to add any more uppercase param names in new code

camsaul commented on modules/drivers/redshift/test/metabase/driver/redshift_test.clj:
```diff
@@ -1,10 +1,15 @@
-            [metabase.test.data.datasets :refer [expect-with-driver]]))
+            [metabase.query-processor :as qp]
+            [metabase.test :as mt]
+            [metabase.test.data.datasets :refer [expect-with-driver]]
```
you've probably found this out already but the linters are going to complain if you import namespaces and then don't use them

camsaul replied:
Another quick heads up:

This is related to the switch from uppercase param-names to backticks in docstrings: we switched from using the [Expectations](https://github.com/clojure-expectations/expectations) library for testing to `clojure.test` last year mostly because of better support in 3rd-party tools. I wrote [a handful of macros](https://github.com/metabase/metabase/blob/master/test/expectations.clj) that expand Expectations-style `expect` tests into `clojure.test` `deftest` ones so we could switch without having to rewrite thousands of tests immediately. 

As with the param names, we've been reworking old tests so they use `deftest` rather than the legacy `expect` in a piecemeal fashion. I usually rework a namespace whenever I add tests to one if it's not already reworked. If you feel like reworking any tests, please feel free to do so. Just make sure to take the time and rework the tests thoughtfully so they follow the spirit of `clojure.test` rather than just macroexpanding the `expect` forms and programatically generating names for them. There are a lot of cases where we can make test code a lot more concise with `clojure.test` so it's worth taking the time to do so.

Either way, whatever you do, don't add any new tests that use `expect` or any of the various other versions we have out in the wild

camsaul commented:
I think this is looking good. To run tests against all the drivers you need to include `[ci all]` or `[ci drivers]` in the commit message. (You can also run tests against a single driver with `[ci redshift]` or the like).

We don't run driver tests automatically because some of the tests like Snowflake can take as much as 20 minutes to run. So for the time being, we're pretty much on the honor system. At any rate, tests against `master` or any `release-*` branch always run the full test suite.

So just make sure when you're writing code that potentially affects multiple drivers to add `[ci drivers]` to the commit message. 

Pro tip: CircleCI uses Docker for everything so with the exception of Redshift, Oracle, and Snowflake, you can run all of the other databases we support locally with the same Docker image we use in CI. If you're fixing a broken test it's a lot faster to do it from the REPL than wait for CI. You can take a look at the config file in `.circleci/config.yml` for the Docker images/env vars we use. Let me know if you need help with that.

camsaul commented on modules/drivers/redshift/test/metabase/driver/redshift_test.clj:
```diff
@@ -1,10 +1,15 @@
-            [metabase.test.data.datasets :refer [expect-with-driver]]))
+            [metabase.query-processor :as qp]
+            [metabase.test :as mt]
+            [metabase.test.data.datasets :refer [expect-with-driver]]
```
I just wrote a long note about how we switched from Expectations to `clojure.test` and accidentally deleted it. So check your email and hopefully it's there so I don't have to type it out again. If not the TL;DR is we switched to `clojure.test` so don't use `expect*` for writing new tests (and bonus points for rewriting old ones to use the new style)

robdaemon replied:
removed use of expect in this test, will apply that going forward too

camsaul commented on modules/drivers/redshift/test/metabase/driver/redshift_test.clj:
```diff
@@ -16,3 +21,47 @@
+   (mt/test-driver
+    :redshift
+    (is (= expected
+           (query->native
```
You can use `mt/mbql-query` shorthand to generate most of this query for you:

```clj
(assoc
  (my/mbql-query users
    {:filter [:or [:= $id [:value 1 {:name "userid"}]]
                  [:= $name [:value "Rafael" {:name "firstname"}]]
                  [:= $name [:value "Robert" {:name "firstname"}]]]
     :limit 2000})
  :info {...})
```

would actually macroexpand to the exact same query you've written. Should save time

camsaul commented:
A few more notes:

It would be good to add more of an end-to-end test that actually converts a query as the FE client would generate it and verifies it has the remark we expect. [This BigQuery test](https://github.com/metabase/metabase/blob/master/modules/drivers/bigquery/test/metabase/driver/bigquery/query_processor_test.clj#L185) is basically what I have in mind. The FE client doesn't generate filter clauses with `[:value ...]` in them -- those wrappers are added in middleware -- so we need to have coverage to make sure the whatever ends up happening, the correct remark gets generated.

Commit:
Code review feedback

Commit:
Setting optional_account_id with site-uuid

Removed timezone test that isn't testing the driver itself, since the
driver doesn't specify a default timezone

Removed deprecated code from the redshift test

Commit:
Field-values handling

Leave the original parameters from the front-end as :user-parameters so
that they are accessible down in the drivers

Transform the original parameters into the field_values in the query
remark

[ci drivers] - to force the drivers to be tested

camsaul reviewed (approved):
looks good to me!
